### PR TITLE
Addition of working assembly based hybrid mPMT geometry.

### PR DESCRIPTION
**Pull request by patrickstowell**
Merge of new realistic PMT geometry based on integration diagrams.
Issues with mPMT hits have been traced back to a rotation and offset issue in the WCPMT2 objects. This has been resolved and now weighted PMT hits are agreeing within 1-2%.

**Comparisons of the 500MeVmum_xdir validation set**
ORIGINAL PMT1 COUNTS [AVGNHIT,NPMT]: 2108.33 19826 
REALISTIC PMT1 COUNTS [AVGNHIT,NPMT]: 2059.1 19746
COMPARISON WEIGHTED HITS OLD VS NEW : 0.106341 0.104279 -> 0.98061

ORIGINAL PMT2 COUNTS [AVGNHIT,NPMT]: 599.71 188613
REALISTIC PMT2 COUNTS [AVGNHIT,NPMT]: 49.13 15504
COMPARISON WEIGHTED HITS OLD VS NEW : 0.003179 0.003168 -> 0.99654

### CODE REVIEW

File: src/WCSimDetectorConfigs.cc
Comment (tdealtry) [inline src/WCSimDetectorConfigs.cc]: Think this is test code that can be removed?

File: src/WCSimDetectorConfigs.cc
Comment (tdealtry) [inline src/WCSimDetectorConfigs.cc]: I presume you're not using this in the realisitc placement code? So can you remove these?
(I guess there could be some weird old geometry code logic reasoning for why these are required)

Comment (tdealtry) [reply src/WCSimDetectorConfigs.cc]: Same with PMTs per cell, etc.

File: src/WCSimDetectorConfigs.cc
Comment (tdealtry) [inline src/WCSimDetectorConfigs.cc:1092]: Again presume can be removed

File: src/WCSimDetectorConfigs.cc
Comment (tdealtry) [inline src/WCSimDetectorConfigs.cc:895]: It'd be very useful to have a `SetHyperK_HybridmPMT_IDOnly_Realistic_Geometry` option too

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: offset for what...?

Actually it'd be nice to comment this a little more to description to these variables (a lot are obvious, some less so)

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:71]:
```suggestion
    // Black tyvek is place on the wall of the ID
    G4VisAttributes* BlackTyvekVis;
```
Relatedly, is the black sheet actually tyvek?

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:78]:
```suggestion
    // Dead space is the space between ID & OD
    G4VisAttributes* DeadSpaceVis;
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Make it clear that ID inner is expected to be 0 (it confused me why an inner radius was required for ID, but makes sense after seeing the value below)

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:87]: Clarify what inner & outer is (think this is obvious, but ID has inner & outer too which makes less sense to me, so should be clear here)

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:100]:
```suggestion
    // White tyvek is placed on the instrumented (inner) side of the OD
    G4VisAttributes* WhiteTyvekVis;
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:108]:
```suggestion
    // Wall tyvek is placed on the inner wall of the water tank
    G4VisAttributes* WallTyvekVis;
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:126]: Explain what the rock shell is

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:119]: Explain what the main water tank is

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: unfinished comment

File: src/WCSimDetectorConfigs.cc
Comment (tdealtry) [inline src/WCSimDetectorConfigs.cc:1049]: Please remove, as you override them in realistic placement

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: The comment suggests you changed it, but you haven't...

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: These sound like `int`s

Also, please explain all these!

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Explain all these too!

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Can you add comments to each colour line, to say what colour it is?
Actually, maybe better would be to collect all the default colours together in a readme somewhere

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: This is actually 2pi too...

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:555]: Can you explain why this might be useful?

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
    // | #F# |
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
    // skip the top left position in the cell, as it is empty (position_high, rotation_low)
```
I like the idea you have with the empty parts having a placeholder, but prefer to have a comment instead of commented out code

Choose your favourite wording & then please repeat it for all similar lines

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
    // Then we repeat but also add mPMT for PMT_hybrid_three_cell
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Do we need this? If we do, can you add what the numbers mean? And remove uncessary lines (e.g. `.q`)

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Not required for the PR, but just reminding you about your reminder

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
      // Simple row only ID pmt
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: unfinished comment

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
    // Now cap the other side
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Please use G4cout, G4cerr & G4endl instead
(There was some mangling when mixing these with std::cout etc in the past)

**Review by tdealtry — CHANGES_REQUESTED** [review]
Thanks for this @patrickstowell Well done tracking down the discrepency!!

I've made (what turns out to be quite a lot of) comments inline, but most are trivial (typos, requests for more comments, request for removing code, etc.)

The biggest one is probably an OD free geometry (definition of this TBD. It maybe makes sense to do things differently to previous, and just have it identical except for the presence of OD PMTs). But after reading the realistic placement file I realise this will be non trivial, so won't hold this PR up for it

Another request is to use `checkOverlaps` & `checkOverlapsPMT`, which are set to true when the code is built with `-DWCSim_Geometry_Overlaps_CHECK=ON`, to turn on/off the Geant4 overlap checking. This will mean that the geometry can be debugged easily, but also it can be constructed quickly when people just need to use it. Again, I won't hold the PR up for this one

A final request is to add the new geometry option to `WCSim.mac` (commented out), along with a comment specifiying which geometry parameters effect the geo & which are ignored. It'd be nice to have that with this PR

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc:187]:
```suggestion
  G4ThreeVector& /*position*/,
```
Unused variable

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Can you remove these, to get rid of the compilation warnings
```
WCSimConstructRealisticPlacement.cc:310:14: warning: declaration of ‘twopi’ shadows a global declaration [-Wshadow]
     G4double twopi = 3.14159*2;
```

```suggestion
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]: Unused variable compilation warning
```suggestion
    //G4double InnerTyvekRadius = config.InnerDetectorOuterRadius;
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
    //int NBlocksUp = config.NBlocksUp;
```
Unused variable compilation warning

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
      std::cout << "HK FD Detector Configuration Table (units: mm)" << std::endl;
```

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
      std::cout << "WhiteTyvekOuterRadius = " << WhiteTyvekOuterRadius << std::endl;
```
Fixing copy-paste error

File: src/WCSimConstructRealisticPlacement.cc
Comment (tdealtry) [inline src/WCSimConstructRealisticPlacement.cc]:
```suggestion
      std::cout << "RockShellRadius = " << RockShellRadius << std::endl;
```
Fixing copy-paste error

**Review by tdealtry — CHANGES_REQUESTED** [review]
Few more triival things to do (noticed upon compiling & running)

**Comment by patrickstowell** [comment]
@tdealtry Actually turning off the ODPMTs is straight forward if all we want to do is not place them but still keep the OD water volume. I can do that before the merge, but it'll need a warning as in that configuration light is still simulated in the OD water volume. To fix that we'd need to set the OD water volume to be a 'non-optical' version of the Water material so GEANT4 knows to kill photons generated in that region.

File: src/WCSimDetectorConfigs.cc
Comment (patrickstowell) [reply src/WCSimDetectorConfigs.cc:1092]: All these horrible legacy variables in DetectorConfigs I am assuming they can go, but I'll need a bit of time to check they don't break anything if they are just left dangling at the values the default SK geometry puts them at. Some seem to. For now I think we should compromise by separating them out in DetectorConfigs and fixing their values with a comment to say they shouldn't do anything. Then I'll spend a bit of time repeating validations with them all removed after the merge.

File: src/WCSimConstructRealisticPlacement.cc
Comment (patrickstowell) [reply src/WCSimConstructRealisticPlacement.cc]: Removed. This was copied over because construct cylinder also mentions an offset.

**Comment by tdealtry** [comment]
Yeah not simulating the light propagation in the OD region sounds like the best thing to do, otherwise we're not saving much CPU time (i.e. if we're only turning the PMTs off).

If this is complicated, can do this in a future PR (just make sure to have the warning about light propagating in the OD)

File: src/WCSimDetectorConfigs.cc
Comment (tdealtry) [reply src/WCSimDetectorConfigs.cc:1092]: Sounds good

**Review by tdealtry — APPROVED** [review]

**Comment by tdealtry** [comment]
Thanks @patrickstowell

**Pull request merged**
